### PR TITLE
start of the @Frames macro

### LINKED ISSUE #331: Extension to RFrames
**Is your feature request related to a problem? Please explain.**
When specifying a relative frame range it would be would be nice to have an option to specify if the reference frame was the previous object's first or last frame. A particular use case is when objects need to be added to the canvas one after another based on their starting frame.

**Describe the solution you'd like**
Add the argument `from_start::Bool` to `RFrames` to indicate the frame range for this object. 
Example:
```
ball = Object(Rframes(5:10, true), (args...)->somefunc())
```

**Describe alternatives you've considered**
A workaround is to create a function to access the starting frame of the desired object in a way like `start(obj)`. The only problem seems to be when the previous object (or a reference to it) is not known.

**Additional context**
Both the methods were suggested by @Wikunia on Javis' zulip [thread](https://julialang.zulipchat.com/#narrow/stream/253420-javis/topic/Extension.20to.20Rframes.3F).

### PR DESCRIPTION
**Pull request by Wikunia**
## PR Checklist

If you are contributing to `Javis.jl`, please make sure you are able to check off each item on this list:

- [x] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [x] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [x] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [x] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [x] Did I properly add test dependencies to the `test` directory (if applicable)?
- [ ] Did I check relevant tutorials that may be affected by changes in this PR?
- [x] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**
Closes #331 

**How did you address these issues with this PR? What methods did you use?**
Based on our discussion in #334 I've created a `@Frames` macro.
One can now write `prev_start()` to refer to the previous starting frame of the last object or action. Similarly with `prev_end`.
These can be used like this in the `@Frames` macro.
```julia
Background(1:70, ground)
red_ball = Object(1:70, (args...) -> object(O, "red"), Point(100, 0))
act!(red_ball, Action(anim_rotate_around(2π, O)))
blue_ball = Object(@Frames(prev_start(),70), (args...) -> object(O, "blue"), Point(200, 80))
act!(blue_ball, Action(anim_rotate_around(2π, 0.0, red_ball)))
act!(blue_ball, Action(@Frames(prev_start(), 10), appear(:fade)))
```

The first parameter of `@Frames` is the start of the frame (same as when just writing a unit range but can be a function like `prev_start()` or something like `prev_start()+10` or whatever)
The second parameter is the length of the object/action.

In the above example it's quite simple such that it can be easily achieved with unit ranges but one might want to do more complicated stuff like 

```julia
objects = [
    Object(@Frames(prev_start()+2, stop=200), (args...) -> circle(O, 10, :fill), points[i]) for i in 1:npoints
]
```
which is now easier to write than previously:
```julia
objects = [
  Object(@Frames(prev_start()+2, stop=200), (args...) -> circle(O, 10, :fill), points[i]) for 
     (frame_start, i) in zip(1:2:(2 * npoints), 1:npoints)
]
```

Additionally it supports
```
startof(obj/act)
endof(obj/act)
```
which are similar to `prev_start` and `prev_end` but one can specify the object or action.

### CODE REVIEW

**Comment by Wikunia** [comment]
@codejaeger you might want to check whether this is enough for your work or whether I missed something.

**Comment by codejaeger** [comment]
Hey @Wikunia , the requirements I was having for specifying actions on the node objects will be met by what you have provided in the PR. I will take a closer look into the code shortly and using the tests that I had prepared in the PR #334  I will be able to point out if anything might seem missing. Those tests might be a bit unreadable so I will mention the cases in verbose that I think should also be fulfilled with this PR.

**Comment by Wikunia** [comment]
I think we might want to add something like `start(object_name)` as well along side `prev_start()` to just have access to the frames of any object.

**Comment by Wikunia** [comment]
Did you do some more tests with this @codejaeger ?

**Comment by codejaeger** [comment]
Hey @Wikunia , yes I tried this PR for my purposes in graph animations. I was primarily looking for a way to provide a default case where the frame range for the next object can be calculated with respect to the starting frame of the previous object/action and an absolute finish frame.  This PR achieves that.

There were a few more utilities that I think will be useful. For example `parent_start` for both objects and actions. If this was available, I would not need to pass an exact frame number to `stop`  and could just do `stop=parent_last()` instead. For objects parent would be the background and for actions they should be the objects. I cannot think of a use case for `prev_start(object)` since if a user has a reference to the other object itself they could reorder creation of the objects to achieve it. We need to think if it would be useful or not.

These are probably the only use cases I can think of right now.

**Comment by Wikunia** [comment]
I think `start(object)` makes sense as the object can be created way before and reordering does not always help as one might want to write `start(object)+300` or something along those lines. Regarding `parent_end` for objects it might be useful for layers or something but I think it's not necessary for background as one can simply use `Background(1:nframes)` and then just write `, stop=nframes`

**Comment by codejaeger** [comment]
Yes agreed, for backgrounds it is `parent_start/end` is not that useful.

**Comment by codejaeger** [comment]
@Wikunia , would it be possible to add just one other frame utility function `prev_action_end` which returns the last frame of the object on which an object is defined. I have a use case where I wanted to add actions serially to different objects one after another. 

This requires the knowledge of the action frames across different objects. I think it would not be possible unless a reference to that action was available. This would be a problem in the implementation since the `start` argument of the macro is executed during rendering, but `prev_action_end` refers to the last action in the last object at that point of time. A possible workaround could be this

```julia
a = Action(RFrames(1:5), appear(:fade))
act!(obj1, a)
act!(obj2, Action(@Frames(prev_action_end(a)+12), appear(:fade)))
```

with a possible implementation for `prev_action_end` being
```julia
function prev_action_end(a)
    return a.frames.frames[end] + PREVIOUS_OBJECT[1].frames.frames[1] - 1
end
```
Do you have an alternative around this?

**Comment by Wikunia** [comment]
Oh you're talking about applying it to different objects. 
I'm not sure I understand it though. Do you mean you want it relative to the the new object or not?

So bascially
```
obj1 = Object(1:100...
obj2 = Object(50:100...
act!(obj1, Action(10:20...
act!(obj2, Action(@Frames(action_end(obj1), 20, ...
```
Should that be the same as
```
act!(obj2, Action(20:39...
```
so 
```
act!(obj2, Action(GFrames(70:89)...
```
? 
Hope it's not too confusing :smile: 

That should work as we compute the objects first as long as `obj1` is defined `obj2` as we compute the frames first for all objects and then for all actions ordered by their object.

**Comment by codejaeger** [comment]
Yes something like that but a little bit more flexible such that the even if more actions were added to `obj1` I would be able to get the frames for `Action(10:20..`  when I specify `@Frames(action_end(obj1)` i.e. get the last frame for the last action defined on the last object at the point of time of writing the code 😆 .

I could formulate this as a different problem of being able to give the animator a handle to the current active frame in the video i.e. the frame last used. This helps provide a sequential style for creating the video without worrying about global frames but the previous specification of the problem makes it more exact to what I need.

**Comment by Wikunia** [comment]
Oh based on when you call the code I missed that point. That will be quite hard to make at this stage. Maybe you can create a new issue for that and at least get this one going into Javis earlier then?

**Comment by codejaeger** [comment]
Yes, I am already working on an alternative but based on the use case below I think it might be useful.
```julia
for i in 1:8
end
```

**Comment by Wikunia** [comment]
And please give more information for the new issue without knowledge of your graph project :wink:

**Comment by codejaeger** [comment]
@Wikunia can this PR be merged? I wanted to extend the `CURRENT_*` for a custom object type.

**Comment by Wikunia** [comment]
Feel free to merge this into your fork for now but yeah want to merge this soon to Javis. Any objections @TheCedarPrince ?

**Comment by Wikunia** [comment]
@Sov-trotter do you mind having a look at the error message here? I'm unsure how this happens and I can't reproduce it locally.

**Comment by Sov-trotter** [comment]
This seems weird. Could you try running the CI again?

**Comment by Wikunia** [comment]
Thanks for having a look. Unfortunately it has the same error as before. 
I realized however that I should only delete the file after the stream was canceled. :wink:

File: test/unit.jl [hunk 1]
```diff
@@ -267,6 +267,24 @@
...
+        a2 = Action(@Frames(startof(a1) + 5, 5), anim_scale(1, 2))
+        act!(obj1, a1)
+        act!(obj1, a2)
+        obj2 = Object(@Frames(prev_start(), 10), (args...) -> 1)
+        obj3 = Object(@Frames(startof(obj2) + 1, 10), (args...) -> 1)
+
+        Javis.preprocess_frames!(demo.objects)
+        @test Javis.get_frames(obj1) == 1:10
+        @test Javis.get_frames(obj2) == 1:10
+        @test Javis.get_frames(obj3) == 2:11
```
Comment (TheCedarPrince) [inline test/unit.jl:285]: I am really baffled that this works but `Javis.get_frames` does not always return frame ranges.

File: src/structs/Frames.jl [hunk 2]
```diff
@@ -95,3 +96,99 @@ function get_frames(parent, elem, glob::GFrames, last_frames::UnitRange; is_firs
...
+function prev_end()
+    if CURRENT_OBJECT_ACTION_TYPE[1] == :Object
+        PREVIOUS_OBJECT[1].frames.frames[end]
+    else
+        PREVIOUS_ACTION[1].frames.frames[end]
+    end
+end
+
+startof(oa::Union{AbstractAction,AbstractObject}) = oa.frames.frames[1]
+endof(oa::Union{AbstractAction,AbstractObject}) = oa.frames.frames[end]
```
Comment (TheCedarPrince) [inline src/structs/Frames.jl:150]: I am confused by the usefulness of these shorthands as right now, I am unable to get frames starts and end values from inputted Objects. Where can this be used?

**Review by TheCedarPrince — COMMENTED** [review]
Hey Ole, I am rather confused by the exact functionality of this PR. In some of my testing, I was trying to reason about where `@Frames` created objects were beginning and ending, and discovered that the following snippet produces different results:

```julia
using Javis

video = Video(600, 400)
Background(1:200, ground)

red_circ = Object(1:90, (args...)->circ("red"))
blue_circ = Object(@Frames(prev_start()+20, 70), (args...)->circ("blue"))
blue_circ = Object(@Frames(prev_start()+20, stop=90), (args...)->circ("blue"))

println(Javis.get_frames(blue_circ))

red_circ = Object(1:90, (args...)->circ("red"))
blue_circ = Object(21:90, (args...)->circ("blue"))
blue_circ = Object(41:90, (args...)->circ("blue"))

println(Javis.get_frames(blue_circ))
```

which gives:

```julia
include("tmp/tmp.jl")

nothing
41:90
```

So somehow, it seems like `@Frames` created objects are not the same as "normally" created objects. Could you explain to me what is happening here? I looked in the macro definition and it seems that you are setting the frame ranges to `nothing`. Why is that? Thanks!

File: examples/follow_path.jl [hunk 3]
```diff
@@ -39,8 +39,11 @@ Background(1:220, ground)
 
 # let the points appear one by one
 objects = [
-    Object(frame_start:200, (args...) -> circle(O, 10, :fill), points[i]) for
-    (frame_start, i) in zip(1:2:(2 * npoints), 1:npoints)
+    Object(
+        @Frames(prev_start(), 200 - prev_start() + 1),
```
[the remark this answers is not in the document]
Comment (Wikunia) [reply examples/follow_path.jl:43]: Yeah I think I implement stop afterwards 😂

**Comment by Wikunia** [comment]
Thanks for the review @TheCedarPrince 
About your first question:
In some cases one can't know the frame range from certain objects when they get created. We have the `preprocess_frames!` function for that.

File: test/Project.toml [hunk 4]
```diff
@@ -2,10 +2,11 @@
 Animations = "27a7e980-b3e6-11e9-2bcd-0b925532e340"
 Gtk = "4c0ca9eb-093a-5379-98c5-f87ac0bbbf44"
 GtkReactive = "27996c0f-39cd-5cc1-a27a-05f136f946b6"
-Interact = "c601a237-2ae4-5e1e-952c-7a85b0c7eef1"
 ImageIO = "82e4d734-157c-48bb-816b-45c225c6df19"
 ImageMagick = "6218d12a-5da1-5696-b52f-db25d2ecc6d1"
 Images = "916415d5-f1e6-5110-898d-aaa5f9f070e0"
+Interact = "c601a237-2ae4-5e1e-952c-7a85b0c7eef1"
+Javis = "78b212ba-a7f9-42d4-b726-60726080707e"
```
[the remark this answers is not in the document]
Comment (Wikunia) [reply test/Project.toml:9]: good catch.

File: test/unit.jl (hunk 1 above)
Comment (Wikunia) [reply test/unit.jl:285]: works because of `Javis.preprocess_frames!(demo.objects)` a few lines above

File: src/structs/Frames.jl (hunk 2 above)
Comment (Wikunia) [reply src/structs/Frames.jl:150]: You can do:
```julia
obj1 = Object(@Frames(prev_start(), 10), (args...) -> 1)
obj2 = Object(2:100,  (args...) -> 1)
obj3 = Object(@Frames(startof(obj2) + 1, 10), (args...) -> 1)
```        
previously one would have to write `obj3` before `obj2` to be able to use `prev_start()` which most often is probably a good idea anyway to have them in order they appear but in certain cases it might be better to specify the object directly.

File: examples/follow_path.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (Wikunia) [reply examples/follow_path.jl]: Is fixed.

File: examples/follow_path.jl [hunk 5]
```diff
@@ -39,8 +39,11 @@ Background(1:220, ground)
...
+        (args...) -> circle(O, 10, :fill),
+        points[i],
+    ) for _ in 1:npoints
```
[the remark this answers is not in the document]
Comment (Wikunia) [reply examples/follow_path.jl]: fixed

File: src/Javis.jl [hunk 6]
```diff
@@ -266,8 +262,11 @@ function render(
         @error "Currently, only gif and mp4 creation is supported. Not a $ext."
     end
 
+    # check if livestream is used and livestream if that's the case
     _livestream(streamconfig, framerate, video.width, video.height, pathname)
 
+    empty_CURRENT_constants()
```
[the remark this answers is not in the document]
Comment (Wikunia) [reply src/Javis.jl:269]: done

**Comment by Wikunia** [comment]
@Sov-trotter any idea why this might happen? https://github.com/Wikunia/Javis.jl/pull/338/checks?check_run_id=3142096015#step:7:422
I'll probably just rerun it later...

**Comment by Sov-trotter** [comment]
MIght be realted to https://trac.ffmpeg.org/ticket/6463#comment:16 ??

**Comment by Wikunia** [comment]
What this PR does which should probably be in a different one is: 
- removing the gifs that are created and streamed to not add two gifs to the test directory which might potentially end up in Javis when one doesn't take care when one commits and pushes to GitHub. I might delete that from this PR.

**Comment by Wikunia** [comment]
@TheCedarPrince do you mind having another look at this i.e resolving conversations and let me know if there is anything I should change.

**Review by TheCedarPrince — APPROVED** [review]
LGTM!
